web/src/HUDGrid.tsx

```diff
@@ -0,0 +1,105 @@
+import React from "react"
+import styled from "styled-components"
+import { Height, Width, ZIndex } from "./constants"
+
+// The HUD UI looks like this:
+//
+// ----------------------------------
+// | top bar                    | S |
+// -----------------------------| i |
+// | resource bar               | d |
+// -----------------------------| e |
+// | main pane                  | b |
+// |                            | a |
+// |                            | r |
+// |                            |   |
+// |                            |   |
+// |                            |   |
+// ----------------------------------
+// | status bar                     |
+// ----------------------------------
+//
+// Our DOM architecture needs to satisfy two competing constraints:
+//
+// 1) "window scroll" needs to scroll the main pane.
+//    Inner-div-scrolls generally have lots of accessibility/UX problems
+//    Nick can tell you hours of stories about this.
+//    https://medium.engineering/the-case-of-the-eternal-blur-ab350b9653ea
+//
+// 2) The topbar / sidebar / statusbar need to visually frame the main pane,
+//    like normal UI panels. Stuff shouldn't get stuck under them.
+//
+// To create this layout, we make the topbar, sidebar, and statusbar position: fixed.
+// Then we create a ~ShadowGrid~ underneath them to preserve margins.
+
+type HUDGridProps = {
+  topBar: React.ReactNode
+  resourceBar: React.ReactNode
+
+  // The main pane
+  children: React.ReactNode
+
+  isSidebarClosed: boolean
+}
+
+let Root = styled.div`
+  display: flex;
+  align-items: stretch;
+  width: 100%;
+`
+
+let TopFrame = styled.div`
+  position: fixed;
+  top: 0;
+  left: 0;
+  right: 0;
+  transition: margin-right 300ms ease;
+  z-index: ${ZIndex.topFrame};
+  margin-right: ${Width.sidebar}px;
+  height: ${Height.topBar}px;
+
+  ${Root}.is-resourcebarvisible & {
+    height: ${Height.topBar + Height.resourceBar}px;
+  }
+
+  ${Root}.is-sidebarclosed & {
+    margin-right: ${Width.sidebarCollapsed}px;
+  }
+`
+
+let MainFrame = styled.div`
+  display: flex;
+  align-items: stretch;
+  width: 100%;
+  padding-bottom: ${Height.statusbar}px;
+  transition: margin-right 300ms ease;
+  padding-top: ${Height.topBar}px;
+  margin-right: ${Width.sidebar}px;
```

Comment:
Is it problematic to have width 100% AND a right- (or left-) margin? Could it ever create a hard-to-interpret situation for some browsers?

(I honestly am not sure. This is a thing I've done a lot but recently started to suspect)

Reply:
ya, i think making it padding-right is probably more correct here